documentation/design_docs/design/plans.adoc

```diff
@@ -0,0 +1,174 @@
+
+* name (this would be the name of the config map)
+
+* displayName (a possibly different name more suitable for displaying
+  to users, with less constraints on allowed chars)
+
+* description
+
+* resources, a list of objects each containing the resource name
+  (currently broker and router are recognised resources) and 'min' and
+  'max' values for the number of pods of that resource an address of
+  this plan would allow. There would also be an 'aggregate' resource,
+  whose 'max' value would limit the total number of pods allowed
+
+* addressPlans, a list of the names of AddressPlan definitions valid
+  for address spaces operating under this plan
+
+* uuid to identify this plan for the service broker
+
+An AddressPlan has the following fields:
+
+* name, unique within the namespace (and by implication generally
+  unique within the cluster also)
+
+* displayName (doesn't need to be unique)
+
+* description
+
+* addressType, indicating the type of address it is associated with
+
+* requiredResources, a list of objects containing the resource name (currently
+  broker and router are recognised resources) and a credits field
+  whose value is a floating point number indicating the amount of that
+  resource an address of this plan would require
+
+The console could then determine both the address-types it could offer
+as well as the plans supported for each type just by reading the
+config maps containing AddressPlan definitions. It could also use the
+provided display_name and description.
+
+The current apporach of parsing tooltips out of the docs will be
+abandoned for now, as the plan configuration will replace most of the
+current usage.
+
+The standard-controller would use both the AddressSpacePlan and
+AddressPlans to determine when to scale up a deployment and to enforce
+limits. The standard-controller would also allocate any queues or
+topics that were to be deployed on a pool of shared broker to one of
+the brokers in that pool. It would mark the address with that
+broker's identity. The agent would use that information to determine which
+broker to create the address on.
+
+TODO: need to determine exactly how the addresses are marked.
+
+Example AddressSpacePlan definition:
+
+{
+    "apiVersion": "enmasse.io/v1",
+    "kind": "AddressSpacePlan",
+    "metadata": {
+        "name": "my-address-space-plan"
+    },
+    "displayName": "a really good plan",
+    "shortDescription": "Demonstrates the syntax",
+    "longDescription": "Insert lengthy discussion of my plan in here...",
+    "uuid": "e47135de-28d5-336d-95ed-5a816cf61658"
+    "resources": [
+        {
+            "name": "broker",
+            "min": 0,
+            "max": 5
+        },
+        {
+            "name": "router",
+            "min": 1,
+            "max": 5
+        },
+        {
+            "name": "aggregate",
+            "max": 6
+        }
+    },
+    "addressPlans":["queue-myplan", "topic-small", "topic-large"]
```

Comment:
Did we agree on making templates optional or separate in the first pass? If optional, we could add a
```
"template": {
  "templateName": "standard-space-infra",
  "templateParameters": [
    {
       "name": "MY_TEMPLATE_PARAMETER",
       "value": "42"
    }
}
```

Reply:
If we envisage perhaps moving it out at a later stage (and overhauling it), my instinct would be to keep it separate.

However as long as it is optional and there are the appropriate warnings about using it, I don't have a problem including it in the plans.

Reply:
My main motivation for including them in the plan would be that it would make converting the current static plan interface to the configmap-backed ones a bit easier.